processing/src/main/java/io/druid/query/extraction/CascadeExtractionFn.java

```diff
@@ -0,0 +1,199 @@
+    }
+    this.chainedExtractionFn = root;  }
+
+  @JsonProperty
+  public ExtractionFn[] getExtractionFns() {
+    return extractionFns;
+  }
+
+  @Override
+  public byte[] getCacheKey() {
+    byte[] cacheKey = new byte[1];
+    cacheKey[0] = CACHE_TYPE_ID;
+
+    return (chainedExtractionFn != null) ? Bytes.concat(cacheKey, chainedExtractionFn.getCacheKey()) : cacheKey;
+  }
+
+  @Override
+  public String apply(Object value) {
+    return (chainedExtractionFn != null) ? chainedExtractionFn.apply(value) : null;
+  }
+
+  @Override
+  public String apply(String value){
+    return (chainedExtractionFn != null) ? chainedExtractionFn.apply(value) : null;
+  }
+
+  @Override
+  public String apply(long value){
+    return (chainedExtractionFn != null) ? chainedExtractionFn.apply(value) : null;
+  }
+
+  @Override
+  public boolean preservesOrdering(){
+    return chainedExtractionFn == null || chainedExtractionFn.preservesOrdering();
+  }
+
+  @Override
+  public ExtractionType getExtractionType(){
+    return (chainedExtractionFn != null) ? chainedExtractionFn.getExtractionType() : null;
+  }
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    CascadeExtractionFn that = (CascadeExtractionFn) o;
+
+    if (!Arrays.equals(extractionFns, that.extractionFns)) {
+      return false;
+    }
+    if (chainedExtractionFn != null && !chainedExtractionFn.equals(that.chainedExtractionFn)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return (chainedExtractionFn != null) ? chainedExtractionFn.hashCode() : 0;
+  }
+
+  @Override
+  public String toString() {
+    return "CascadeExtractionFn{" +
+        "extractionFns=[" + ((chainedExtractionFn != null) ? chainedExtractionFn.toString() : "") + "]}";
+  }
+
+  private class ChainedExtractionFn {
+    private final ExtractionFn fn;
+    private final ChainedExtractionFn child;
+
+    public ChainedExtractionFn(ExtractionFn fn, ChainedExtractionFn child) {
+      this.fn = fn;
+      this.child = child;
+    }
+
+    public byte[] getCacheKey() {
+      byte[] fnCacheKey = fn.getCacheKey();
+
+      return (child != null) ? Bytes.concat(fnCacheKey, child.getCacheKey()) : fnCacheKey;
+    }
+
+    public String apply(Object value) {
+      return fn.apply((child != null) ? child.apply(value) : value);
+    }
+
+    public String apply(String value){
+      return fn.apply((child != null) ? child.apply(value) : value);
+    }
+
+    public String apply(long value){
+      return fn.apply((child != null) ? child.apply(value) : value);
+    }
+
+    public boolean preservesOrdering(){
+      boolean childPreservesOrdering = (child == null) || child.preservesOrdering();
+      return fn.preservesOrdering() && childPreservesOrdering;
+    }
+
+    public ExtractionType getExtractionType(){
+      return (child != null)
```

Comment:
I think this set of conditions is a bit hard to read, can you change it to something like:

if (child != null && child.getExtractionType() == ExtractionType.MANY_TO_ONE) {
  return ExtractionType.MANY_TO_ONE;
} else {
  return fn.getExtractionType(); 
}